### PR TITLE
Merge branch 'main' into feature/proof-of-stake

### PR DESCRIPTION
**Pull request by tholonious**
Add additional event manager type to emit state change events to file to be consumed by another job

Remove unused function

Add operation type

Update index size

Update how struct size is calculated

Update tests

Resolve conflicts

Remove protobuf, rely on deso encoders for v1

Update state syncer encoding to use deso encoders

Update encoding format

Add db flush events, update encoding types

Update to state syncer

Fix config flags

Fix tests

Fix tests

Update logging

Updates to state syncer

Merge conflicts

Create state change encoder

Update state change syncer

Updates to state syncer encoder

Clean up state change file handling

Simplify state syncer, use utxobundle, add documentation

Fix tests

More test fixes

Update tests, run go fmt

Go fmt

Add additional entry type support to state syncer

Additional entry type support

Updates to state change encoder

Merge conflicts with balance model, tweaks to mempool handler

Fix mempool sync condition

Updates to state syncer index file format

Updates to core

Add additional entry type for balance entry

Additional entry type support

Updates to state change metadata

Add state change metadata

Fix dao coin state change metadata

Updates to limit order state change metadata

Update block sync procedure

Updates to ss core

Updates to state syncer routine

Address LN feedback

Fix state syncer flush error handling

Updates to backend config

Update state syncer decoder

Minor state syncer bug fixes

Fix test compile errors

Resolve merge conflicts

Merge conflicts

Fix test compiler issues

Updates to testing

Update isMempool signatures

Track mempool transactions

Fix mempool flush

Address feedback, fix CI

Formatting

Update mempool sync routine to reduce number of re-flushes of mempool

Updates to state change metadata

Add mempool lock to prevent concurrent map reads, update transaction fees

Updates to extra data decoder

ln/add-transactor-as-affected-pub-key (#675)

* ln/add-transactor-as-affected-pub-key

* address DH feedback

---------

Co-authored-by: Lazy Nina <>

Updates to get pkid routine

Updates to sync procedure

Update ancestral record decoding

Update state change entry struct and encoders

Merge main, resolve conflicts

Address formatting and print feedback

Merge pull request #488 from superzordon/z/state-syncer-golang

Merge branch 'main' into feature/proof-of-stake

Adapt PoS code to work with state syncer

### CODE REVIEW

**Comment by tholonious** [comment]
Current dependencies on/for this PR:
* feature/proof-of-stake
  * **PR #807** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/807?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/807?utm_source=stack-comment).

File: lib/block_view_flush.go [hunk 1]
```diff
@@ -1438,7 +1410,7 @@ func (bav *UtxoView) _flushPostAssociationEntriesToDbWithTxn(txn *badger.Txn, bl
 		} else {
 			// If the AssociationEntry has isDeleted=false then we
 			// put the corresponding mappings for it into the db.
-			if err := DBPutPostAssociationWithTxn(txn, bav.Snapshot, &associationEntry, blockHeight); err != nil {
+			if err := DBPutPostAssociationWithTxn(txn, bav.Snapshot, &associationEntry, blockHeight, bav.EventManager); err != nil {
```
Comment (lazynina) [inline lib/block_view_flush.go:1413]: Just a note: we'll need to update the flush logic to support the event manager for all the new PoS entries.

File: lib/block_view_stake.go [hunk 2]
```diff
@@ -984,15 +988,15 @@ func DBDeleteStakeEntryWithTxn(
 
 	// Delete StakeEntry from PrefixStakeByValidatorByStaker.
 	stakeByValidatorAndStakerKey := DBKeyForStakeByValidatorAndStaker(validatorPKID, stakerPKID)
-	if err := DBDeleteWithTxn(txn, snap, stakeByValidatorAndStakerKey); err != nil {
+	if err := DBDeleteWithTxn(txn, snap, stakeByValidatorAndStakerKey, eventManager, isDeleted); err != nil {
```
Comment (lazynina) [inline lib/block_view_stake.go:991]: oh you did it already! nice!!

File: lib/block_view_types.go [hunk 3]
```diff
@@ -110,13 +110,23 @@ const (
...
+	// EncoderTypeDeSoBalanceEntry represents a balance of DeSo for a particular public key.
+	EncoderTypeDeSoBalanceEntry EncoderType = 42
+	// EncoderTypeBlock represents a block in the blockchain, including all transactions in said block.
+	EncoderTypeBlock EncoderType = 43
+	// EncoderTypeTxn represents a transaction in the blockchain.
+	EncoderTypeTxn              EncoderType = 44
+	EncoderTypeValidatorEntry   EncoderType = 45
+	EncoderTypeStakeEntry       EncoderType = 46
+	EncoderTypeLockedStakeEntry EncoderType = 47
+	EncoderTypeEpochEntry       EncoderType = 48
```
Comment (lazynina) [inline lib/block_view_types.go:126]: g000d catch here.

File: scripts/nodes/n2 [hunk 4]
```diff
@@ -1,17 +0,0 @@
-DATE=`date +%Y%m%d-%H:%M:%S-%N`
```
Comment (lazynina) [inline scripts/nodes/n2:1]: Crap we weren't supposed to delete these in main...

**Review by lazynina — APPROVED** [review]

**Review by superzordon — APPROVED** [review]
LGTM 👍

**Pull request merged**